source/extensions/tracers/opencensus/opencensus_tracer_impl.cc

```diff
@@ -0,0 +1,171 @@
+                              SystemTime start_time) override;
+  void setSampled(bool sampled) override;
+
+private:
+  // TODO: Do we take the sampling decision from StartSpan, or use the sampler
+  // configured in TraceConfig?
+  ::opencensus::trace::Sampler* GetSampler(bool is_sampled) {
+    static ::opencensus::trace::AlwaysSampler always_sampler;
+    static ::opencensus::trace::NeverSampler never_sampler;
+    if (is_sampled) {
+      return &always_sampler;
+    }
+    return &never_sampler;
+  }
+
+  ::opencensus::trace::Span span_;
+  const envoy::config::trace::v2::OpenCensusConfig* oc_config_;
+};
+
+Span::Span(const Tracing::Config& config,
+           const envoy::config::trace::v2::OpenCensusConfig* oc_config,
+           Http::HeaderMap& /*request_headers*/, const std::string& operation_name,
+           SystemTime /*start_time*/, const Tracing::Decision tracing_decision)
+    : span_(::opencensus::trace::Span::StartSpan(
+          operation_name, /*parent=*/nullptr,
+          ::opencensus::trace::StartSpanOptions(GetSampler(tracing_decision.traced)))),
+      oc_config_(oc_config) {
+  span_.AddAttribute("OperationName", config.operationName() == Tracing::OperationName::Ingress
+                                          ? "Ingress"
+                                          : "Egress");
+}
+
+Span::Span(const envoy::config::trace::v2::OpenCensusConfig* oc_config,
+           ::opencensus::trace::Span&& span)
+    : span_(std::move(span)), oc_config_(oc_config) {}
+
+void Span::setOperation(const std::string& operation) {
+  span_.AddAnnotation("setOperation", {{"operation", operation}});
+}
+
+void Span::setTag(const std::string& name, const std::string& value) {
+  span_.AddAttribute(name, value);
+}
+
+void Span::finishSpan() { span_.End(); }
+
+void Span::injectContext(Http::HeaderMap& request_headers) {
+  if (oc_config_->propagate_cloud_trace_context()) {
+    request_headers.addCopy(
+        Http::LowerCaseString{"x-cloud-trace-context"},
+        ::opencensus::trace::propagation::ToCloudTraceContextHeader(span_.context()));
+  }
+  if (oc_config_->propagate_trace_context()) {
+    request_headers.addCopy(Http::LowerCaseString{"traceparent"},
+                            ::opencensus::trace::propagation::ToTraceParentHeader(span_.context()));
+  }
+  if (oc_config_->propagate_grpc_trace_bin()) {
+    request_headers.addCopy(
+        Http::LowerCaseString{"grpc-trace-bin"},
+        ::opencensus::trace::propagation::ToGrpcTraceBinHeader(span_.context()));
+  }
+}
+
+Tracing::SpanPtr Span::spawnChild(const Tracing::Config& /*config*/, const std::string& name,
+                                  SystemTime /*start_time*/) {
+  span_.AddAnnotation("spawnChild");
+  return std::make_unique<Span>(oc_config_,
+                                ::opencensus::trace::Span::StartSpan(name, /*parent=*/&span_));
+}
+
+void Span::setSampled(bool sampled) { span_.AddAnnotation("setSampled", {{"sampled", sampled}}); }
```

Comment:
Does this prevent the span from being reported (if sampled is false)? This is called (for example) if the request is associated with a health check, so those requests are not sampled.

Reply:
At this stage, no. OpenCensus only allows sampling decisions to be made at `StartSpan()` time. We're discussing "abandon span" in https://github.com/census-instrumentation/opencensus-specs/issues/218

The best I can do for now is turn the `setSampled` call into an annotation.